src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java

```diff
@@ -265,4 +323,171 @@ public void onError(IOException e, HttpURLConnection uc) throws IOException {
 
     };
 
+    /*package*/ static void checkConnectionValidity(String apiUri, @NonNull TaskListener listener,
+                                                    StandardCredentials credentials,
+                                                    GitHub github)
+            throws IOException {
+        synchronized (githubs) {
+            Map<GitHub,Void> hubs = checked.get(listener);
+            if (hubs != null && hubs.containsKey(github)) {
+                // only check if not already in use
+                return;
+            }
+            if (hubs == null) {
+                hubs = new WeakHashMap<>();
+                checked.put(listener, hubs);
+            }
+            hubs.put(github, null);
+        }
+        if (credentials != null && !github.isCredentialValid()) {
+            String message = String.format("Invalid scan credentials %s to connect to %s, skipping",
+                    CredentialsNameProvider.name(credentials), apiUri == null ? GitHubSCMSource.GITHUB_URL : apiUri);
+            throw new AbortException(message);
+        }
+        if (!github.isAnonymous()) {
+            assert credentials != null;
+            listener.getLogger().println(GitHubConsoleNote.create(
+                    System.currentTimeMillis(),
+                    String.format("Connecting to %s using %s",
+                    apiUri == null ? GitHubSCMSource.GITHUB_URL : apiUri,
+                    CredentialsNameProvider.name(credentials))
+            ));
+        } else {
+            listener.getLogger().println(GitHubConsoleNote.create(System.currentTimeMillis(), String.format(
+                    "Connecting to %s with no credentials, anonymous access",
+                    apiUri == null ? GitHubSCMSource.GITHUB_URL : apiUri
+            )));
+        }
+    }
+
+    /*package*/
+    static void checkApiRateLimit(@NonNull TaskListener listener, GitHub github)
+            throws IOException, InterruptedException {
+        boolean check = true;
+        while (check) {
+            check = false;
+            long start = System.currentTimeMillis();
+            GHRateLimit rateLimit = github.rateLimit();
+            long rateLimitResetMillis = rateLimit.getResetDate().getTime() - start;
+            double resetProgress = rateLimitResetMillis / MILLIS_PER_HOUR;
+            // the buffer is how much we want to avoid using to cover unplanned over-use
+            int buffer = Math.max(15, rateLimit.limit / 20);
+            // the burst is how much we want to allow for speedier response outside of the throttle
+            int burst = rateLimit.limit < 1000 ? Math.max(5, rateLimit.limit / 10) : Math.max(200, rateLimit.limit / 5);
+            // the ideal is how much remaining we should have (after a burst)
+            int ideal = (int) ((rateLimit.limit - buffer - burst) * resetProgress) + buffer;
```

Comment:
So the idea is a linear slope from the start of the hour to the end of the hour. We want to keep at least `buffer` unused. We want to start the ideal line, not from the top of the rate limit, but from a bit down so that we have room to burst at the top of the hour. This is just linear maths.